service/app/contract_events_handler.go

```diff
@@ -0,0 +1,320 @@
+	}
+
+	txScript, err := flow_helpers.ParseCadenceTemplate(
+		OPEN_SCRIPT,
+		&flow_helpers.CadenceTemplateVars{
+			CollectibleNFTName:    contract.Name,
+			CollectibleNFTAddress: contract.Address.String(),
+		},
+	)
+	if err != nil {
+		return err // rollback
+	}
+
+	t, err := transactions.NewTransactionWithDistributionID(OPEN_SCRIPT, txScript, arguments, distribution.ID)
+	if err != nil {
+		return err // rollback
+	}
+
+	if err := t.Save(ev.db); err != nil {
+		return err // rollback
+	}
+
+	ev.eventLogger.Info("Pack open transaction created")
+	return nil
+}
+
+func (ev *EventHandler) HandleOpenedEvent(ctx context.Context, event *flow.Event, pack *Pack, distribution *Distribution) error {
+
+	// Make sure the pack is in correct state
+	if err := pack.Open(); err != nil {
+		return fmt.Errorf("error while handling %s: %w", OPENED, err)
+	}
+
+	// Update the pack in database
+	if err := UpdatePack(ev.db, pack); err != nil {
+		return err // rollback
+	}
+	return nil
+}
+
+func (ev *EventHandler) PollByEventName(ctx context.Context, wg *sync.WaitGroup, cpc *CirculatingPackContract, cpcCursor *CirculatingPackContractBlockCursor, eventName string, begin uint64, end uint64) error {
+
+	arr, err := ev.flowClient.GetEventsForHeightRange(ctx, client.EventRangeQuery{
+		Type:        cpc.EventName(eventName),
+		StartHeight: begin,
+		EndHeight:   end,
+	})
+
+	if err != nil {
+		return fmt.Errorf("error querying events for height range. err=%w", err)
+	}
+
+	for _, be := range arr {
+		for _, e := range be.Events {
+			ev.eventLogger = ev.eventLogger.WithFields(log.Fields{"eventType": e.Type, "eventID": e.ID()})
+
+			ev.eventLogger.Debug("Handling event")
+
+			evtValueMap := flow_helpers.EventValuesToMap(e)
+
+			packFlowIDCadence, ok := evtValueMap["id"]
+			if !ok {
+				return fmt.Errorf("could not read 'id' from event %s", e)
+			}
+
+			packFlowID, err := common.FlowIDFromCadence(packFlowIDCadence)
+			if err != nil {
+				return err // rollback
+			}
+
+			addressLocation := AddressLocation{
+				Name:    cpc.Name,
+				Address: cpc.Address,
+			}
+			pack, err := GetPackByContractAndFlowID(ev.db, addressLocation, packFlowID)
+			if err != nil {
+				return fmt.Errorf("error retrieving pack from db err=%w", err) // rollback
+			}
+
+			distribution, err := GetDistributionSmall(ev.db, pack.DistributionID)
+			if err != nil {
+				return fmt.Errorf("error retrieving distribution from db err=%w", err)
+			}
+
+			ev.eventLogger = ev.eventLogger.WithFields(log.Fields{
+				"distID":     distribution.ID,
+				"distFlowID": distribution.FlowID,
+				"packID":     pack.ID,
+				"packFlowID": pack.FlowID,
+			})
+
+			ev.eventLogger.Info("handling event...")
+
+			if err := ev.HandleEvent(ctx, eventName, &e, pack, distribution); err != nil {
+				if errors.Is(err, &common.InvalidPackStateError{}) {
+					ev.eventLogger.Warnf("InvalidPackState error, continuing. err:%s", err.Error())
```

Comment:
Created a new error type, `InvalidPackStateError` so we can determine when to continue event processing / when to break and try again